cpp/src/rolling/rolling_detail.cuh

```diff
@@ -1105,6 +1171,13 @@ struct rolling_window_launcher {
     auto gather_map = create_collect_gather_map(
       offsets->view(), per_row_mapping->view(), preceding_begin, stream, mr);
 
+    // If gather_map collects null elements, and null_policy == EXCLUDE,
+    // those elements must be filtered out, and offsets recomputed.
+    auto null_handling = static_cast<collect_list_aggregation*>(agg.get())->_null_handling;
+    if (null_handling == null_policy::EXCLUDE && input.has_nulls()) {
+      purge_null_entries(input, gather_map, offsets, stream, mr);
+    }
```

Comment:
Is the gather map here guaranteed to include all elements of the input table?
If so, we don't need the complex null check to short-circuit purge_null_entries - any null in input column is guaranteed to affect the output.

Reply:
> Is the gather map here guaranteed to include all elements of the input table?

It is astute of you to notice this. The condition might be rare (especially for calls from Spark's window functions), because each input row's window is likely to touch/overlap with its neighbours'. But it is possible that not all input rows might be selected.

But `rolling_window()` has a generalized version that takes a `preceding_col` and `following_col`. This version of the API allows each row's window to be of arbitrary size, causing rows to be left out, especially at the ends.

Reply:
P.S. Is `has_nulls()` complex? Should this be avoided if possible?

Reply:
I meant that the check within `purge_null_entries` could be avoided if all elements are guaranteed to be a part of the aggregation. This is not the case, so I'm resolving the convo :)